examples/07_end_to_end/seismic_waveform.ipynb

```diff
@@ -0,0 +1,1334 @@
+      "2012-12-14 10:42:37.624998 TA.E42A..BHN:TA.E42A..BHN | 2012-12-14T10:42:33.474998Z - 2012-12-14T10:42:37.624998Z | 40.0 Hz, 167 samples\n",
+      "2012-12-14 10:42:37.624998 TA.E42A..BHZ:TA.E42A..BHZ | 2012-12-14T10:42:33.474998Z - 2012-12-14T10:42:37.624998Z | 40.0 Hz, 167 samples\n",
+      "2012-12-14 10:42:41.599998 TA.E42A..BHE:TA.E42A..BHE | 2012-12-14T10:42:37.649998Z - 2012-12-14T10:42:41.599998Z | 40.0 Hz, 159 samples\n",
+      "2012-12-14 10:42:41.599998 TA.E42A..BHN:TA.E42A..BHN | 2012-12-14T10:42:37.649998Z - 2012-12-14T10:42:41.599998Z | 40.0 Hz, 159 samples\n",
+      "2012-12-14 10:42:41.599998 TA.E42A..BHZ:TA.E42A..BHZ | 2012-12-14T10:42:37.649998Z - 2012-12-14T10:42:41.599998Z | 40.0 Hz, 159 samples\n"
+     ]
+    }
+   ],
+   "source": [
+    "client = Client(\"IRIS\")\n",
+    "t = UTCDateTime(\"2012-12-14T10:36:01.6Z\")\n",
+    "starttime = (t + 300).datetime\n",
+    "endtime = (t + 400).datetime\n",
+    "@csp.graph\n",
+    "def graph():\n",
+    "    basket = fdsn_waveforms(client, \"TA\", \"E42A\", \"*\", \"BH?\", timedelta(seconds=4))\n",
+    "    for k, v in basket.items():\n",
+    "        csp.print(k, v)\n",
+    "\n",
+    "out = csp.run(graph, starttime=starttime, endtime=endtime)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "85bfc73c-db25-44fa-a0a5-02ac22ca0b5f",
+   "metadata": {},
+   "source": [
+    "## Real-time Data via Seedlink\n",
+    "In this section, we build a real time adapter to live waveform data. It is straightforward to apply the filtering and triggering logic from the previous section to these outputs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96fc1fee-0d20-45f5-8724-b6905dcbcd54",
+   "metadata": {},
+   "source": [
+    "### Real-time Adapter Definition"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "id": "4808fc0e-53a1-4677-aa28-8542ce8f31f5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import threading\n",
+    "from fnmatch import fnmatch\n",
+    "from obspy.clients.seedlink.client.seedlinkconnection import SeedLinkConnection\n",
+    "from obspy.clients.seedlink.client.slstate import SLState\n",
+    "from obspy.clients.seedlink.slpacket import SLPacket\n",
+    "from urllib.parse import urlparse\n",
+    "\n",
+    "from csp.impl.adaptermanager import AdapterManagerImpl\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "\n",
+    "class EasySeedLinkAdapter:\n",
+    "    def __init__(self, server_url: str):\n",
+    "        parsed_url = urlparse(server_url, scheme=\"seedlink\")\n",
+    "        self.server_hostname = parsed_url.hostname\n",
+    "        self.server_port = parsed_url.port or 18000\n",
+    "        self.conn = SeedLinkConnection()\n",
```

Comment:
may want to move this into the impl and just pass server_url to the impl
better demonstration that the wiring time object should really just hold the necessary args